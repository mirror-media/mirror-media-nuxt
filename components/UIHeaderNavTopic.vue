<template>
  <section class="header-nav-topic">
    <div class="container">
      <div class="topic-container">
        <nuxt-link
          v-for="topic in topics"
          :key="topic.id"
          class="topic topic--normal"
          :to="`/topic/${topic.id}`"
          @click.native="emitGA(`topic ${topic.name}`)"
        >
          <h2>{{ topic.name }}</h2>
        </nuxt-link>
      </div>

      <nuxt-link
        to="/section/topic"
        class="topic"
        @click.native="emitGA('topic 更多')"
      >
        <h2>更多</h2>
      </nuxt-link>

      <div class="sub-brand-container">
        <a
          href="https://voice.mirrorfiction.com/"
          target="_blank"
          @click="emitGA('section mirrorvoice')"
        >
          <img src="~/assets/mirrorvoice.png" alt="鏡好聽" />
        </a>
        <a
          href="https://www.mirrorfiction.com/"
          target="_blank"
          @click="emitGA('section mirrorfiction')"
        >
          <img src="~/assets/mirrorfiction.png" alt="鏡文學" />
        </a>
        <a
          href="https://www.readr.tw/"
          target="_blank"
          @click="emitGA('section readr')"
        >
          <img src="~/assets/readrlogo-gary.png" alt="READr 讀+" />
        </a>
      </div>
    </div>
  </section>
</template>

<script>
export default {
  name: 'UIHeaderNavTopic',
  props: {
    topics: {
      type: Array,
      required: true,
      default: () => [],
    },
  },
  methods: {
    emitGA(eventLabel) {
      this.$emit('sendGA', {
        eventCategory: 'header',
        eventAction: 'click',
        eventLabel,
      })
    },
  },
}
</script>

<style lang="scss" scoped>
.header-nav-topic {
  display: none;
  @include media-breakpoint-up(xl) {
    display: block;
    background-color: #fff;
    color: rgba(#000, 0.5);
    text-align: center;
  }
  .container {
    max-width: 1024px;
    height: 45px;
    overflow: hidden;
    display: flex;
    margin-left: auto;
    margin-right: auto;
    user-select: none;
  }
  .topic-container {
    display: flex;
    flex-wrap: wrap;
    max-width: 582px;
  }
  .topic {
    min-width: 90px;
    height: 100%;
    padding-left: 8px;
    padding-right: 8px;
    line-height: 42px;
    cursor: pointer;
    &--normal {
      border-bottom: 3px solid #000;
      flex: 0 0 auto;
      &:hover {
        color: #fff;
        background-color: #000;
      }
    }
  }
  .sub-brand-container {
    display: flex;
    align-items: center;
    flex: 0 0 auto;
    & a {
      padding-left: 10px;
      padding-right: 10px;
      flex: 0 0 auto;
      cursor: pointer;
    }
    & img {
      height: 30px;
      width: auto;
    }
  }
}
</style>
