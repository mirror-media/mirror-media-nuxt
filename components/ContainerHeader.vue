<template>
  <header>
    <button type="button" class="menu-icon" @click="handleClickMenuIcon" />
    <transition name="slide">
      <UISidebar
        v-if="isSidebar"
        :topics="topics"
        :sections="sections"
        :partners="partners"
        :subBrands="subBrandLinks"
        :others="otherLinks"
        :socialMedia="socialMediaLinks"
        @close="handleSidebarClose"
        @sendGA="handleSendGA"
      />
    </transition>

    <div class="header-search">
      <UISearchBarWrapper :options="options" @sendGA="handleSendGA" />
      <UIOthersList
        class="others-list"
        :links="otherLinks"
        eventCategory="header"
        @sendGA="handleSendGA"
      />
    </div>

    <nav class="header-nav">
      <UIHeaderNavSection
        :sections="sections"
        :partners="partners"
        @sendGA="handleSendGA"
      />
      <UIHeaderNavTopic :topics="topics" @sendGA="handleSendGA" />
    </nav>
  </header>
</template>

<script>
import { mapGetters } from 'vuex'

import UISidebar from './UISidebar.vue'
import UISearchBarWrapper from './UISearchBarWrapper.vue'
import UIOthersList from './UIOthersList.vue'
import UIHeaderNavSection from './UIHeaderNavSection.vue'
import UIHeaderNavTopic from './UIHeaderNavTopic.vue'

import {
  SUB_BRAND_LINKS,
  SOCIAL_MEDIA_LINKS,
  OTHER_LINKS,
} from '~/constants/index'

export default {
  name: 'ContainerHeader',
  components: {
    UISidebar,
    UISearchBarWrapper,
    UIOthersList,
    UIHeaderNavSection,
    UIHeaderNavTopic,
  },
  data() {
    return {
      isSidebar: false,
      defaultOption: { title: '全部類別' },
    }
  },
  computed: {
    ...mapGetters({
      sections: 'sections/displayedSections',
      partners: 'partners/displayedPartners',
      topics: 'topics/displayedTopics',
    }),
    options() {
      const sections = this.sections.filter(
        (section) => section.name !== 'videohub'
      )
      return [this.defaultOption, ...sections]
    },
    otherLinks() {
      return this.transformObjIntoArray(OTHER_LINKS)
    },
    socialMediaLinks() {
      return this.transformObjIntoArray(SOCIAL_MEDIA_LINKS)
    },
    subBrandLinks() {
      return this.transformObjIntoArray(SUB_BRAND_LINKS)
    },
  },
  methods: {
    handleClickMenuIcon() {
      this.openSidebar()
      this.sendHeaderGA('menu open')
    },
    handleSidebarClose() {
      this.closeSidebar()
      this.sendHeaderGA('menu close')
    },
    openSidebar() {
      this.isSidebar = true
    },
    closeSidebar() {
      this.isSidebar = false
    },
    sendHeaderGA(eventLabel, eventAction = 'click') {
      this.$ga.event({
        eventCategory: 'header',
        eventAction,
        eventLabel,
      })
    },
    handleSendGA(param = {}) {
      this.$ga.event(param)
    },
    transformObjIntoArray(obj) {
      return Object.values(obj)
    },
  },
}
</script>

<style lang="scss" scoped>
header {
  background-color: #f5f5f5;
}
.menu-icon {
  width: 24px;
  height: 40px;
  background-image: url(~assets/hamburger@2x.png);
  background-size: 24px;
  background-position: center;
  background-repeat: no-repeat;
  cursor: pointer;
  user-select: none;
  @include media-breakpoint-up(xl) {
    display: none;
  }
}
.header-search {
  display: flex;
  align-items: center;
  z-index: 149;
  position: relative;
}
.others-list {
  display: none;
  @include media-breakpoint-up(xl) {
    display: block;
  }
}
.header-nav {
  box-shadow: 0 2px 1px rgba(#000, 0.2);
  position: relative;
  z-index: 99;
  @include media-breakpoint-up(xl) {
    box-shadow: 0 0 5px #ccc;
  }
}
.slide-enter-active,
.slide-leave-active {
  transition: transform 0.45s;
}
.slide-enter,
.slide-leave-to {
  transform: translateX(-100%);
}
</style>
