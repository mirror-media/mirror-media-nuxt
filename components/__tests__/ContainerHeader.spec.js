import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import ContainerHeader from '../ContainerHeader.vue'
import UIHeaderNavSection from '../UIHeaderNavSection.vue'
import UIHeaderNavTopic from '../UIHeaderNavTopic.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Vuex state computed', () => {
  test('get the proper data', () => {
    const store = new Vuex.Store({
      state: {
        sections: { items: [{ name: 'news' }] },
        topics: { items: [{ name: '網紅星勢力' }] },
        partners: { items: [{ name: 'healthnews' }] },
      },
    })
    const wrapper = shallowMount(ContainerHeader, { store, localVue })

    expect(wrapper.vm.sections).toEqual([{ name: 'news' }])
    expect(wrapper.vm.topics).toEqual([{ name: '網紅星勢力' }])
    expect(wrapper.vm.partners).toEqual([{ name: 'healthnews' }])
  })

  test('get an empty array when the data is undefined', () => {
    const store = new Vuex.Store({
      state: {
        sections: {},
        topics: {},
        partners: {},
      },
    })
    const wrapper = shallowMount(ContainerHeader, { store, localVue })

    expect(wrapper.vm.sections).toEqual([])
    expect(wrapper.vm.topics).toEqual([])
    expect(wrapper.vm.partners).toEqual([])
  })
})

describe('handleSendGA method', () => {
  const store = new Vuex.Store({
    state: {
      sections: {},
      topics: {},
      partners: {},
    },
  })

  test('throw error when the argument is invalid', () => {
    const wrapper = shallowMount(ContainerHeader, { store, localVue })

    expect(() => {
      wrapper.vm.handleSendGA()
    }).toThrow(wrapper.vm.gaError)

    const invalidArgs = {
      eventCategoryInvalid: 'header',
      eventAction: 'click',
      eventLabel: 'section news',
    }
    expect(() => {
      wrapper.vm.handleSendGA(invalidArgs)
    }).toThrow(wrapper.vm.gaError)
  })

  test('call $ga method when UIHeaderNavTopic.vue emits sendGA', () => {
    const $ga = {
      event: jest.fn(),
    }
    const wrapper = shallowMount(ContainerHeader, {
      mocks: {
        $ga,
      },
      store,
      localVue,
    })
    const gaArgs = {
      eventCategory: 'header',
      eventAction: 'click',
      eventLabel: 'topic 人造地獄',
    }
    wrapper.find(UIHeaderNavTopic).vm.$emit('sendGA', gaArgs)
    expect($ga.event).toHaveBeenCalledWith(gaArgs)
  })

  test('call $ga method when UIHeaderNavSection.vue emits sendGA', () => {
    const $ga = {
      event: jest.fn(),
    }
    const wrapper = shallowMount(ContainerHeader, {
      mocks: {
        $ga,
      },
      store,
      localVue,
    })
    const gaArgs = {
      eventCategory: 'header',
      eventAction: 'click',
      eventLabel: 'section news',
    }
    wrapper.find(UIHeaderNavSection).vm.$emit('sendGA', gaArgs)
    expect($ga.event).toHaveBeenCalledWith(gaArgs)
  })
})

describe('markup', () => {
  test('has a header tag', () => {
    const store = new Vuex.Store({
      state: {
        sections: {},
        topics: {},
        partners: {},
      },
    })
    const wrapper = shallowMount(ContainerHeader, { store, localVue })

    expect(wrapper.contains('header')).toBe(true)
  })
})
