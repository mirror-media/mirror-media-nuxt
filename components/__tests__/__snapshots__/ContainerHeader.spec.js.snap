// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup render correctly 1`] = `
<header>
  <uisearchbar-stub />
   
  <nav
    class="header-nav"
  >
    <uiheadernavsection-stub
      partners=""
      sections=""
    />
     
    <uiheadernavtopic-stub
      topics=""
    />
  </nav>
</header>
`;
